dm/dm/worker/subtask.go

```diff
@@ -552,7 +552,10 @@ func (st *SubTask) markResultCanceled() bool {
 func (st *SubTask) Result() *pb.ProcessResult {
 	st.RLock()
 	defer st.RUnlock()
-	return st.result
+	tempProcessResult, _ := st.result.Marshal()
+	newProcessResult := &pb.ProcessResult{}
+	_ = newProcessResult.Unmarshal(tempProcessResult)
```

Comment:
ditto. Or:
```suggestion
newProcessResult.Unmarshal(tempProcessResult)
```
Is it necessary to decode and encode it again? What about deep copying the object?
```
*a = *b // a, b are pointers
```

Reply:
Actually, it is necessary to decode and encode it, because Result() originally returns a pointer without the lock, which may cause data race when multiple functions are executing concurrently and calling Result(). In this case, deep copying would still return a pointer without the lock, which does not solve the problem. Here decoding and encoding again return a new copy of the process result, so that when multiple functions are calling Result() concurrently and using the returned pointer to modify things, they would be modifying the new copy, which has no effect on the original pointer, thus avoiding the potential data race.

Reply:
Modified using deep copy.

Reply:
_ = newProcessResult.Unmarshal(tempProcessResult)
Here I cannot delete _ =, because 'make check' would fail and report error return value is not checked. Since it is guaranteed here that Unmarshal() will not return an error, so we can just ignore it.